src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -1298,15 +1298,35 @@ private static Attribute NewAliasAttribute(AttributeAst ast)
 
         private static Attribute NewValidateSetAttribute(AttributeAst ast)
         {
+            ValidateSetAttribute result;
             var cvv = new ConstantValueVisitor { AttributeArgument = true };
-            var args = new string[ast.PositionalArguments.Count];
-            for (int i = 0; i < ast.PositionalArguments.Count; i++)
+
+            // 'ValidateSet([CustomGeneratorType], IgnoreCase=$false)' is supported in scripts.
+            if (ast.PositionalArguments.Count == 1 && ast.PositionalArguments[0] is TypeExpressionAst generatorTypeAst)
             {
-                args[i] = _attrArgToStringConverter.Target(_attrArgToStringConverter,
-                    ast.PositionalArguments[i].Accept(cvv));
+                if (TypeResolver.TryResolveType(generatorTypeAst.TypeName.FullName, out Type generatorType))
+                {
+                    result = new ValidateSetAttribute(generatorType);
+                }
+                else
+                {
+                    throw InterpreterError.NewInterpreterException(ast, typeof(RuntimeException), ast.Extent,
+                        "CustomValidValuesGeneratorTypeNotFound", ParserStrings.CustomValidValuesGeneratorTypeNotFound, generatorTypeAst.TypeName.FullName, typeof(System.Management.Automation.IValidateSetValuesGenerator).FullName);
```

Comment:
If we didn't have this special case code, you would have gotten `TypeNotFound`, which is exactly the problem - it has nothing to do with `IValidateSetValuesGenerator`. Besides, if you don't implement the interface, that error is reported when constructing the attribute.